### PR TITLE
Add audit log for password changed events

### PR DESCRIPTION
**Pull request by pkong-ds**
ref DEV-1468

## Events
```
password.primary.reset
password.primary.changed
password.secondary.change
password.primary.force_changed
password.secondary.force_changed
```

### activity_type in portal
image

### Preview 1 - Forgot password
https://github.com/user-attachments/assets/9ca13d4f-00be-4efa-8f2a-a2c356455669

### Preview 2 - Force update
https://github.com/user-attachments/assets/e93224f5-8a86-4bc6-93c6-6a072380166c

### Preview 3 - Settings > change primary password

https://github.com/user-attachments/assets/be5af9fc-ab5d-4bee-8a44-cdcc24c86cd1

### Preview 4 - Settings > change secondary password
https://github.com/user-attachments/assets/a83b64f3-352a-4c5a-941d-1fdb6606d7d0

## Additional context
Modified previous PR - https://github.com/authgear/authgear-server/pull/4722

### CODE REVIEW

**Comment by pkong-ds** [comment]
`npm run ensure-reproducible-build` failed. Fixed after rebase.

**Comment by pkong-ds** [comment]
Actually have 2 questions

1. if the audit log failed, should we cancel entire db transaction and rollback?
2. Do we want some test cases to ensure some `persistAuditLog` fn is called?

File: pkg/api/event/nonblocking/user_force_update_password_changed.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+)
+
+const (
+	UserForceUpdatePasswordChanged event.Type = "user.force_update.password_changed"
+)
+
+type UserForceUpdatePasswordChangedEventPayload struct {
+	UserRef   model.UserRef   `json:"-" resolve:"user"`
+	UserModel model.User      `json:"user"`
+	Sessions  []model.Session `json:"sessions"`
```
Comment (louischan-oursky) [inline pkg/api/event/nonblocking/user_force_update_password_changed.go:16]: Remove sessions

File: pkg/api/event/nonblocking/user_forgot_password_password_changed.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+)
+
+const (
+	UserForgotPasswordPasswordChanged event.Type = "user.forgot_password.password_changed"
+)
+
+type UserForgotPasswordPasswordChangedEventPayload struct {
+	UserRef   model.UserRef   `json:"-" resolve:"user"`
+	UserModel model.User      `json:"user"`
+	Sessions  []model.Session `json:"sessions"`
```
Comment (louischan-oursky) [inline pkg/api/event/nonblocking/user_forgot_password_password_changed.go:15]: Remove sessions.

File: pkg/api/event/nonblocking/user_settings_primary_password_changed.go [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+)
+
+const (
+	UserSettingsPrimaryPasswordChanged event.Type = "user.settings.primary_password_changed"
+)
+
+type UserSettingsPrimaryPasswordChangedEventPayload struct {
+	UserRef   model.UserRef   `json:"-" resolve:"user"`
+	UserModel model.User      `json:"user"`
+	Sessions  []model.Session `json:"sessions"`
```
Comment (louischan-oursky) [inline pkg/api/event/nonblocking/user_settings_primary_password_changed.go:15]: Remove sessions.

File: pkg/api/event/nonblocking/user_settings_secondary_password_changed.go [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+)
+
+const (
+	UserSettingsSecondaryPasswordChanged event.Type = "user.settings.secondary_password_changed"
+)
+
+type UserSettingsSecondaryPasswordChangedEventPayload struct {
+	UserRef   model.UserRef   `json:"-" resolve:"user"`
+	UserModel model.User      `json:"user"`
+	Sessions  []model.Session `json:"sessions"`
```
Comment (louischan-oursky) [inline pkg/api/event/nonblocking/user_settings_secondary_password_changed.go:15]: Remove sessions.

File: pkg/lib/feature/forgotpassword/service.go [hunk 5]
```diff
@@ -450,7 +451,15 @@ func (s *Service) ResetPasswordByEndUser(code string, newPassword string) error
...
-	return s.resetPassword(target, state, newPassword, CodeChannelUnknown)
+	err = s.resetPassword(target, state, newPassword, CodeChannelUnknown)
+	if err != nil {
+		return err
+	}
+	err = s.persistAuditLog(state.UserID)
+	if err != nil {
+		return err
+	}
+	return nil
```
Comment (louischan-oursky) [inline pkg/lib/feature/forgotpassword/service.go:462]: Make `resetPassword` do `persistAuditLog`

File: pkg/lib/feature/forgotpassword/service.go [hunk 6]
```diff
@@ -580,3 +597,14 @@ func (s *Service) setPassword(options *SetPasswordOptions) (err error) {
 func (s *Service) ChangePasswordByAdmin(options *SetPasswordOptions) error {
 	return s.setPassword(options)
 }
+
+func (s *Service) persistAuditLog(userID string) (err error) {
+	err = s.Events.DispatchEventImmediately(&nonblocking.UserForgotPasswordPasswordChangedEventPayload{
```
Comment (louischan-oursky) [inline pkg/lib/feature/forgotpassword/service.go:602]: Use `DispatchEventOnCommit` instead.

File: pkg/lib/authenticationflow/declarative/node_login_flow_change_password.go [hunk 7]
```diff
@@ -70,6 +72,12 @@ func (n *NodeLoginFlowChangePassword) ReactTo(ctx context.Context, deps *authflo
 			return authflow.NewNodeSimple(&NodeSentinel{}), nil
 		}
 
+		// Changed
+		err = n.persistAuditLog(deps, newInfo.UserID)
+		if err != nil {
+			return nil, err
+		}
+
```
Comment (louischan-oursky) [inline pkg/lib/authenticationflow/declarative/node_login_flow_change_password.go:80]: We do not dispatch event in ReactTo. Effects should be done in OnCommitEffect.

File: pkg/lib/accountmanagement/service_authenticator.go [hunk 8]
```diff
@@ -34,6 +35,18 @@ func (s *Service) ChangePrimaryPassword(resolvedSession session.ResolvedSession,
...
+		}
+		// Changed password successfully
+		userID := resolvedSession.GetAuthenticationInfo().UserID
+		err = s.Events.DispatchEventOnCommit(&nonblocking.UserSettingsPrimaryPasswordChangedEventPayload{
+			UserRef: model.UserRef{
+				Meta: model.Meta{
+					ID: userID,
+				},
+			},
+		})
```
Comment (louischan-oursky) [inline pkg/lib/accountmanagement/service_authenticator.go:49]: Move the event dispatch to `s.changePassword`. To ensure every password change has a event.

File: pkg/lib/accountmanagement/service_authenticator.go [hunk 9]
```diff
@@ -104,6 +117,18 @@ func (s *Service) ChangeSecondaryPassword(resolvedSession session.ResolvedSessio
...
+		}
+		// Changed password successfully
+		userID := resolvedSession.GetAuthenticationInfo().UserID
+		err = s.Events.DispatchEventOnCommit(&nonblocking.UserSettingsSecondaryPasswordChangedEventPayload{
+			UserRef: model.UserRef{
+				Meta: model.Meta{
+					ID: userID,
+				},
+			},
+		})
```
Comment (louischan-oursky) [inline pkg/lib/accountmanagement/service_authenticator.go:131]: Move the event dispatch to `s.changePassword`. To ensure every password change has a event.

File: pkg/api/event/nonblocking/user_force_update_password_changed.go [hunk 10]
```diff
@@ -0,0 +1,49 @@
+package nonblocking
+
+import (
+	"github.com/authgear/authgear-server/pkg/api/event"
+	"github.com/authgear/authgear-server/pkg/api/model"
+)
+
+const (
```
Comment (louischan-oursky) [inline pkg/api/event/nonblocking/user_force_update_password_changed.go:8]: Change the list of the new events to

```
password.primary.reset
password.primary.changed
password.secondary.changed
password.primary.force_changed
password.secondary.force_changed
```

Document these new events in specs/event.md

**Review by louischan-oursky — CHANGES_REQUESTED** [review]
I will take care of the change request myself.

**Comment by pkong-ds** [comment]
Thanks daic 🫡

File: pkg/api/event/nonblocking/user_force_update_password_changed.go (hunk 10 above)
Comment (pkong-ds) [reply pkg/api/event/nonblocking/user_force_update_password_changed.go:8]: Question - do we support force-change 2ndary password?

**Review by louischan-oursky — APPROVED** [review]

**Pull request merged**